### PR TITLE
[PatternLang] Convert PatternGrouper to do pre-order, non-recursive analysis

### PR DESCRIPTION
**Pull request by mbrookhart**
Fixes https://github.com/apache/incubator-tvm/issues/5647

@comaniac @masahi Thanks!

### CODE REVIEW

File: src/relay/ir/dataflow_matcher.cc [hunk 1]
```diff
@@ -432,26 +432,39 @@ class PatternGrouper : protected MixedModeVisitor {
...
+   * If we traverse the graph in post-order, we can run into situtations where a small subgraph will
+   * match the pattern. Due to options like AltPattern, a larger subgraph with more nodes later in
+   * the graph may also match the pattern. With post-order traversal, we mark the smaller subgraph
+   * as matched and fail to catch the larger subgraph. This problem is fixed by using pre-order
+   * traversal.
+   */
+  void VisitExprs() {
+    std::unordered_set<Expr, ObjectHash, ObjectEqual> pre_partitioned;
+    for (size_t i = matcher_->expr_graph_.topological_order_.size(); i != 0; --i) {
+      size_t index = i - 1;
```
Comment (masahi) [inline src/relay/ir/dataflow_matcher.cc:456]: Having both `i` and `index` seems redundant.

Comment (mbrookhart) [reply src/relay/ir/dataflow_matcher.cc:456]: This is something I was too lazy to debug.

If I iterate over something of size 4 with the constraint `i > 0`, I get 3 iterations with i = 3, 2, 1

If I use the line you gave me, I get 5 iterations, with i = 3, 2, 1, 0, 18446744073709551615 as the size_t underflows.

I can't figure out _why_ that line screws up

Comment (mbrookhart) [reply src/relay/ir/dataflow_matcher.cc:456]: Ah, nvm, I see. Of course i >=0, it's a size_t. I need to use an int.

Comment (masahi) [reply src/relay/ir/dataflow_matcher.cc]: Ah sorry you are right. What I suggested may get a warning from compiler I think (size_t to int conversion). Moreover, `matcher_->expr_graph_.topological_order_.size() - 1` would wrap around if `matcher_->expr_graph_.topological_order_` is empty.  

So I think your original code is safer. Can you undo the last commit? Sorry about this.

Comment (mbrookhart) [reply src/relay/ir/dataflow_matcher.cc:456]: No worries! Undid it.

Comment (masahi) [reply src/relay/ir/dataflow_matcher.cc:456]: Thanks. Grepping the code base with `i - 1` finds many instances of this, and your original code seems to be the solution :)

**Comment by masahi** [comment]
Thanks @mbrookhart @comaniac

**Pull request merged**